### DISCUSSION TITLE
Support Plural Forms

### DISCUSSION THREAD

**Discussion by TheAwiteb** [welpo/tabi#273] [General]
Hello @welpo, I want to support plural forms in Tabi, but this needs to be done in the `i18n` directory. So do you mind?

My language is Arabic, and it has a few plural forms (unlike English, which only has two). I can help you with the translations if you want.

> [!NOTE]
> I don't know if aouther languages have more than two plural forms, I'll add to Arabic only and keep the rest as they are.

For example, the word "minute" in Arabic has three plural forms:

- دقيقة (singular)
- دقيقتان (dual)
- دقائق (few)
- دقيقة (many)

So, I will add the following to the translations file:

```toml
one_read_minute = "دقيقة" # Only one minute
two_read_minute = "دقيقتان" # Two minutes
few_read_minute = "دقائق" # Numbers from 3 to 10
many_read_minute = "دقيقة" # Numbers higher than 10
```

And the same thing for posts and tags, I think that's only.

**Comment by TheAwiteb** [comment]
Resources:
- https://pontoon.mozilla.org/ar/firefox/all-resources/?search=minutes&string=193946
- https://arabeyes.org/Plural_Forms
- https://docs.translatehouse.org/projects/localization-guide/en/latest/l10n/pluralforms.html

**Comment by welpo** [comment]
As you can see from [your link](https://docs.translatehouse.org/projects/localization-guide/en/latest/l10n/pluralforms.html), pluralisation is not easy. Russian, for example, does this:

```
    0 - Plural
    1 - Singular
2,3,4 - Dual
 5-20 - Plural
   21 - Singular  
22-24 - Dual
25-30 - Plural 
```

You can imagine how complex the logic would get if we need to manually implement pluralisation for all languages.ç

---

I think we would need a macro to decide, based on the language and the number, the translation `key`. For example (Python pseudocode):

```py
# This macro needs two values:
#     - language
#     - number

if language == "ar":
    if number == 1:
        key_prefix = "one_"
    elif number == 2:
        key_prefix = "two_"
    # etc.

# At the end, if key_prefix is still unset, use a default which should match current behaviour.
if not key_prefix:
    if number == 1:
        # Would match to "singular_result", for example, which would be the new name for the current `result` string.
        key_prefix = "singular_"
    else:
        key_prefix = "plural_"  # 0 and > 1

return key_prefix
```

Then use it (not sure this would work):

```tera
key_value = "results"
{{ set key = key_prefix ~ key_value }}
{{ macros_translate::translate(key=key, default="words", language_strings=language_strings) }}"
```

The first line determines the word we're looking for. The second line joins the prefix (the job of the new macro) with that word. Finally, we used the join `key` with th existing translation macro.

---

Something like this would be my first approach.

The new macro (say, `p11n` or `pluralization`) could start small, and slowly grow in complexity.

In summary:

I don't like the idea of doing proper pluralisation, but I do like the end result. Many i18n strings would need to change (say, to include `$MINUTES` and have the variable replaced later) and I don't think the implementation would not be easy.

Is this the approach you had in mind?

**Reply by TheAwiteb** [reply]
> Is this the approach you had in mind?

Of course not, I thought it's easy, also for Arabic I realized it's hard to do it with Tera, I think we need to use a additional script, and I don't know for real if this are possible with zoal/tera or not

**Reply by welpo** [reply]
It can be done. I'll push a proof of concept soon.

**Reply by welpo** [reply]
It's in #276
